### ISSUE TITLE
syscall: Getrlimit test failures in docker on armv7 / aarch64

### ISSUE THREAD

**Issue by Ikke** [golang/go#38604]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.13.10 linux/arm64
</pre>

### Does this issue reproduce with the latest release?

Yes, 1.14.2 still failes

### What operating system and processor architecture are you using (`go env`)?

Alpine Linux in docker on Alpine Linux on armv7 / aarch64

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/home/buildozer/.cache/go-build"
GOENV="/home/buildozer/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/buildozer/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_arm64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build704765024=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Build golang and run the test suite in a docker container on either armv7 or aarch64

See our [APKBUILD](https://gitlab.alpinelinux.org/alpine/aports/-/blob/12258714d17879e248800807664efaab7ab3390a/community/go/APKBUILD) for details

### What did you expect to see?

Test suite pass

### What did you see instead?

<pre>
 --- FAIL: TestRlimit (0.00s)
    syscall_unix_test.go:332: Getrlimit: save failed: operation not permitted
FAIL
FAIL	syscall	0.029s
</pre>

[Build log](https://gist.github.com/Ikke/ac22c195b7afe00ff074e171b71bd670)

This doesn't happen when we build / run the test suite in a privileged container.

**Comment by Ikke** [comment]
It runs fine on other arches (x86, x86_64, ppc64le, s390x)

**Comment by andybons** [comment]
Does the following program print the resource limits or also fail with an error?

```
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <sys/resource.h>

int main() {
  struct rlimit lim;
  if (getrlimit(RLIMIT_NOFILE, &lim) == 0) {
    printf(
        "limit = %llu\n"
        "hard limit = %llu\n",
        lim.rlim_cur, lim.rlim_max);
  } else {
    fprintf(stderr, "%s\n", strerror(errno));
  }
  return 0;
}
```

**Comment by Ikke** [comment]
That works, yes:

```
~ $ cat test_getrlimit.c                                                                                               
#include <errno.h>                                                                                                     
#include <stdio.h>                                                                                                     
#include <string.h>                                                                                                    
#include <sys/resource.h>                                                                                              
                                                                                                                       
int main() {                                                                                                           
  struct rlimit lim;                                                                                                   
  if (getrlimit(RLIMIT_NOFILE, &lim) == 0) {                                                                           
    printf(                                                                                                            
        "limit = %llu\n"                                                                                               
        "hard limit = %llu\n",                                                                                         
        lim.rlim_cur, lim.rlim_max);                                                                                   
  } else {                                                                                                             
    fprintf(stderr, "%s\n", strerror(errno));                                                                          
  }                                                                                                                    
  return 0;                                                                                                            
}                                                                                                                      
                                                                                                                       
~ $ gcc test_getrlimit.c -o test_getrlimit                                                                             
~ $ ./test_getrlimit                                                                                                   
limit = 1048576                                                                                                        
hard limit = 1048576
```

**Comment by andybons** [comment]
@ianlancetaylor any thoughts?

**Comment by ianlancetaylor** [comment]
Can you run both the C program and the Go program under `strace -f`, and see exactly which system calls they are using for `getrlimit`?  Thanks.

**Comment by Ikke** [comment]
This is the golang test that is failing (I show the last part of the output of the process that is calling getrlimit)
<details>
<summary><code>~ $ PATH=$PWD/bin:$PATH strace -f -s520 -o testrlimit.strace  go test -v syscall -run TestRlimit</code></summary>
<pre>
4822  write(1, "=== RUN   TestRlimit\n", 21 <unfinished ...>
4822  <... write resumed>)              = 21
4822  futex(0x40000aa148, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
4822  <... futex resumed>)              = 0
4822  getrlimit(RLIMIT_NOFILE,  <unfinished ...>
4822  <... getrlimit resumed>0x400007ced8) = -1 EPERM (Operation not permitted)
4822  write(1, "--- FAIL: TestRlimit (0.00s)\n", 29 <unfinished ...>
4822  <... write resumed>)              = 29
4822  write(1, "    syscall_unix_test.go:332: Getrlimit: save failed: operation not permitted\n", 78 <unfinished ...>
4822  <... write resumed>)              = 78
4822  futex(0x40000aa148, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
4822  <... futex resumed>)              = 1
4822  write(1, "FAIL\n", 5 <unfinished ...>
4822  <... write resumed>)              = 5
4822  write(3, "# test log\n", 11 <unfinished ...>
4822  <... write resumed>)              = 11
4822  close(3 <unfinished ...>
4822  <... close resumed>)              = 0
4822  exit_group(1 <unfinished ...>
4822  <... exit_group resumed>)         = ?
4822  +++ exited with 1 +++
</pre>
</details>

This is the test program provided by @andybons 

<details>
<summary><code>~ $  strace -f -s 512 -o test_getrlimit.strace ./test_getrlimit</code></summary>
<pre>
4843  execve("./test_getrlimit", ["./test_getrlimit"], 0xffffe8cd4d28 /* 7 vars */) = 0
4843  set_tid_address(0xffffb99220c4)   = 4843
4843  mprotect(0xaaaabafcb000, 4096, PROT_READ) = 0
4843  prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=1024*1024, rlim_max=1024*1024}) = 0
4843  ioctl(1, TIOCGWINSZ, {ws_row=64, ws_col=239, ws_xpixel=0, ws_ypixel=0}) = 0
4843  writev(1, [{iov_base="limit = 1048576", iov_len=15}, {iov_base="\n", iov_len=1}], 2) = 16
4843  writev(1, [{iov_base="hard limit = 1048576", iov_len=20}, {iov_base="\n", iov_len=1}], 2) = 21
4843  exit_group(0)                     = ?
4843  +++ exited with 0 +++
</pre>
</details>

So golang is calling getrlimit, which gets permission denied, while the test program calls prlimit64, which succeeds.

**Comment by ianlancetaylor** [comment]
Thanks for the info.  Can you see if https://golang.org/cl/230339 fixes the problem for you?

**Comment by Ikke** [comment]
@ianlancetaylor Yes, I can confirm this patch fixes the issue, all tests succeed now.

**Comment by tianon** [comment]
For the curious, the root cause of this denial is a bug in `libseccomp` where it wrongly presumed `getrlimit` didn't exist on these ARM arches: https://github.com/seccomp/libseccomp/issues/234 (apparently fixed in `libseccomp` 2.4.4).
